internal/api/auth_hooks.go

```diff
@@ -0,0 +1,164 @@
+package api
+
+import (
+	"encoding/json"
+	"errors"
+	"net/url"
+
+	"fmt"
+	"github.com/gofrs/uuid"
+	"github.com/supabase/gotrue/internal/conf"
+	"github.com/supabase/gotrue/internal/storage"
+	"regexp"
+	"strings"
+)
+
+type HookType string
+
+const (
+	PostgresHook HookType = "pg-functions"
+	HTTPHook     HookType = "http"
+)
+
+const (
+	MFAHookRejection = "reject"
+	MFAHookContinue  = "continue"
+)
+
+type AuthHook struct {
+	ExtensibilityPointConfiguration conf.ExtensibilityPointConfiguration
+	payload                         []byte
+	hookType                        HookType
+	event                           string
+	db                              *storage.Connection
+}
+
+// Hook Events
+const (
+	MFAVerificationEvent = "auth.mfa_verfication"
+)
+
+type HookErrorResponse struct {
+	ErrorMessage string `json:"error_message"`
+	ErrorCode    string `json:"error_code"`
+	RetryAfter   bool   `json:"retry_after"`
+}
```

Comment:
In general I would prefer if these are scoped under one `error` subobject instead. So if the developer wants to return an error they can just send back:

```json
{
  "error": {
    "code": "code",
    "message": "message"
  }
}```

Is the RetryAfter necessary for this initial hook? I don't see it as such and not sure how that would work with the sync hooks honestly.